tools/bitfield_gen.md

```diff
@@ -0,0 +1,672 @@
+0x0e`, then the next larger mask will be used. This means values 0 to 13 are
+using 4 bit tags, 14 and 15 are using 8 bit tags, and all 8 bit tag values must
+have at least the bits 1, 2, and 3 set.
+
+Since there is no larger mask size after 8 in the example, the mask value there
+is not important. However, the generator expects the mask of each larger level
+to have at least all bits set that the masks of all smaller sizes have set, so
+we chose the same value `0x0e` in the example. If we had another level, e.g. 16
+bits wide afterwards, we might have chosen `0b11110` (0x1e) for the mask value
+at 8 bit.
+
+The sub mask inclusion means that the generator can check mask sizes
+consecutively. I.e. if `tag_value & 0xe != 0xe`, we always know it is a 4 bit
+tag, no matter how many other tag sizes are declared.
+
+The tag mask value declared for each alternative in the tagged union must fit
+the mask for the size of the tag field in the block. Remember that with `mask 4
+ 0x0e` the `0x0e` bit pattern means that size 4 does *not* apply, and instead
+the next larger size applies. So in the example, because `content_cap` has a
+4-bit tag field, `content_cap` must have a tag value where the bits `0x0e` are
+not all set (e.g. 1 is a legal value). Conversely, `null_cap` has an 8-bit tag,
+and so at least the bits `0x0e` must all be set in the tag value (e.g. 14 or 15
+works, as does 255, but not 1).
+
+Masks and tag slices are mutually exclusive, i.e. only one of these can be used
+for any particular tagged union.
+
+### Raw Value Access
+
+The C representation of blocks, wether in a tagged union or not, is a struct
+containing an array of the architecture word type (`uint32` or `uint64`) of the
+length computed from the number and width of fields in the block. Given that the
+memory layout is fully specified it is possible to operate on these raw values
+directly without breaking the semantics of the generated functions.
+
+This is usually neither recommended nor necessary, and for code undergoing
+formal verification strongly discouraged, because it introduces additional
+verification overhead.
+
+However, one special case is occasionally convenient: when initialising an area
+of memory with 0, and reinterpreting that memory as a bitfield structure, we are
+guaranteed to get a value where all fields are 0. Similarly, returning an all-0
+struct from a function instead of calling the constructor with only 0 parameters
+may be acceptable, although one should check the generated assembly if the
+compiler was not able to optimise the constructor away -- if it does then the
+constructor form is clearer and easier to use in proofs.
+
+The tool will not access padding fields, but is not guaranteed to preserve them
+when copying. Writing/reading padding fields is not guaranteed to be stable.
+
+### Architecture Parameters and Canonical Pointers
+
+The tool supports architectures with 32 and 64 bit word sizes. The command
+
+```bf_gen
+base 32
+```
```

Comment:
It might help to say where this command goes and what it's scope is. This is probably connected to unwritten sections involving the `.bf` file extension and how the bitfield generator is used.

Reply:
Good point. I should make sure that is actually true, but I think you can have it anywhere in the spec.

Reply:
So it turns out I was wrong and you can have multiple base commands and everything will work just fine. The base command covers all subsequent blocks.

Reply:
Have fixed that now and removed the "exactly one" condition later in the restrictions section as well.